### PR TITLE
Add Fast Image Processor for vilt

### PR DESCRIPTION
**Pull request by devxaitist**
# What does this PR do?
Add ViltImageProcessorFast implementation for faster image processing using PyTorch. (#36978)

This PR adds a fast image processor for the Vilt model that leverages PyTorch and torchvision functions instead of PIL/numpy. The implementation improves performance by using tensor operations and enabling GPU processing.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@yonigozlan

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    resample = PILImageResampling.BICUBIC
+    image_mean = IMAGENET_STANDARD_MEAN
+    image_std = IMAGENET_STANDARD_STD
+    size = {"shortest_edge": 384}
+    do_resize = True
+    do_rescale = True
+    do_normalize = True
+    size_divisor = 32
+    do_pad = True
+    model_input_names = ["pixel_values", "pixel_mask"]
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:82]: I think you should have this for this processor:
```suggestion
    default_to_square = False
    model_input_names = ["pixel_values", "pixel_mask"]
```

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+                new_heights = int(new_heights + 0.5)
+                new_widths = int(new_widths + 0.5)
+                # Make dimensions divisible by size_divisor
+                if size_divisor is not None:
+                    new_heights = new_heights // size_divisor * size_divisor
+                    new_widths = new_widths // size_divisor * size_divisor
+
+                # Resize the image
+                stacked_images = F.resize(stacked_images, [new_heights, new_widths], interpolation=interpolation)
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:143]: Let's override `self.resize` and put that in here

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+                    # Create pixel mask (1 for valid pixels, 0 for padding)
+                    pixel_mask = mask_template.clone()
+                    pixel_mask[: original_size[0], : original_size[1]].fill_(1)
+                else:
+                    padded_image = image
+                    pixel_mask = torch.ones(max_size, dtype=torch.int64, device=image.device)
+
+                padded_images.append(padded_image)
+                pixel_masks.append(pixel_mask)
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:195]: Same here, let's put the bulk of the logic in a separate function, and group by shape before padding batched images

File: tests/models/vilt/test_image_processing_vilt.py [hunk 4]
```diff
@@ -141,13 +145,14 @@ def image_processor_dict(self):
...
+            image_processing = image_processing_class(**self.image_processor_dict)
+            self.assertTrue(hasattr(image_processing, "image_mean"))
+            self.assertTrue(hasattr(image_processing, "image_std"))
+            self.assertTrue(hasattr(image_processing, "do_normalize"))
+            self.assertTrue(hasattr(image_processing, "do_resize"))
+            self.assertTrue(hasattr(image_processing, "size"))
+            self.assertTrue(hasattr(image_processing, "size_divisor"))
 
     def test_image_processor_from_dict_with_kwargs(self):
         image_processor = self.image_processing_class.from_dict(self.image_processor_dict)
```
Comment (yonigozlan) [inline tests/models/vilt/test_image_processing_vilt.py:158]: This needs the same modification as the test above

**Review by yonigozlan — COMMENTED** [review]
Hi @devxaitist , thanks for contributing, Looks great! Some things to modifiy to make the processor cleaner

**Comment by devxaitist** [comment]
Hi @yonigozlan, I've updated the branch to reflect the latest changes in the main branch. When you have a moment, I'd appreciate your review on the updated PR. Thank you!

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                stacked_images, do_rescale, rescale_factor, do_normalize, image_mean, image_std
+            )
+            processed_images_grouped[shape] = stacked_images
+
+        processed_images = reorder_images(processed_images_grouped, grouped_images_index)
+
+        # Handle padding if required
+        data = {}
+        if do_pad:
+            data = self._pad_batch(processed_images, return_tensors)
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:142]: let's have the pad function return the images and mask, and defined the data dict here

**Review by yonigozlan — COMMENTED** [review]
Thanks for iterating @devxaitist ! Still changes to be done, also it would be great to override the test_slow_fast_equivalence functions in the test to also compare the padding masks.

**Comment by devxaitist** [comment]
Thank you so much for your valuable feedback, @yonigozlan ! I've implemented all the requested changes, including overriding the test_slow_fast_equivalence functions to compare padding masks.

I greatly appreciate your time and guidance throughout this process. This is my first open-source contribution, and your feedback has been incredibly helpful for my learning. 

When you have a moment, I'd be grateful if you could review these latest changes. Thank you again for your patience and support!

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        Returns:
+            `tuple`: Tuple containing padded images and pixel masks.
+        """
+        # Calculate global maximum dimensions across all images
+        max_size = get_max_height_width(images)
+
+        # Group images by shape before padding
+        grouped_images, grouped_images_index = group_images_by_shape(images)
+        processed_grouped = {}
+
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:218]:
```suggestion
        processed_images= {}
        processed_masks= {}
```

Maybe easier to do that then used the reorder_images function for both masks and images

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            else:
+                padded_images = stacked_images
+                pixel_masks = torch.ones(
+                    (stacked_images.shape[0], max_size[0], max_size[1]),
+                    dtype=torch.int64,
+                    device=stacked_images.device,
+                )
+
+            # Store processed group
+            processed_grouped[shape] = (padded_images, pixel_masks)
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:246]:
```suggestion
            processed_images[shape] = padded_images
            processed_masks[shape] = pixel_masks
```

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+        # Reorder images back to original order
+        padded_images = []
+        pixel_masks = []
+
+        for _, (group_key, position) in grouped_images_index.items():
+            padded_images.append(processed_grouped[group_key][0][position])
+            pixel_masks.append(processed_grouped[group_key][1][position])
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:254]:
```suggestion
        padded_images= reorder_images(processed_images, grouped_images_index)
        pixel_masks= reorder_images(processed_masks, grouped_images_index)
```
Let me know if something like this works!

File: src/transformers/models/vilt/image_processing_vilt_fast.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    # This generated class can be used as a starting point for the fast image processor.
+    # if the image processor is only used for simple augmentations, such as resizing, center cropping, rescaling, or normalizing,
+    # only the default values should be set in the class.
+    # If the image processor requires more complex augmentations, methods from BaseImageProcessorFast can be overridden.
+    # In most cases, only the `_preprocess` method should be overridden.
+
+    # For an example of a fast image processor requiring more complex augmentations, see `LlavaNextImageProcessorFast`.
+
+    # Default values should be checked against the slow image processor
+    # None values left after checking can be removed
```
Comment (yonigozlan) [inline src/transformers/models/vilt/image_processing_vilt_fast.py:72]:
```suggestion
```

**Review by yonigozlan — COMMENTED** [review]
Thanks for iterating @devxaitist looks much better! I suggested a small improvement in the padding function, other than that LGTM!
You'll also have to rebase on main, and change the way the docstrings are handled, as this was recently updated. You'll need to replace the `@add_start_docstrings` with `@auto_docstring` from `...utils`, and move the docstrings of the custom args to the `ViltFastImageProcessorKwargs` class. You can look at other fast image processors to see how it's done.
Thanks again for the great work!

**Review by yonigozlan — APPROVED** [review]
Very nice, LGTM!
I'm glad to hear you enjoyed working on this PR, and congrats on your first open-source contribution! 🤗
